README.rst
================================
Semantic Label Converter [slcpy]
================================

.. image:: https://img.shields.io/github/v/release/SMLC-NYSBC/Semantic_Label_Creator
        :target: https://img.shields.io/github/v/release/SMLC-NYSBC/Semantic_Label_Creator

.. image:: https://github.com/SMLC-NYSBC/Semantic_Label_Creator/actions/workflows/python-publish_PyPi.yml/badge.svg
        :target: https://github.com/SMLC-NYSBC/Semantic_Label_Creator/actions/workflows/python-publish_PyPi.yml

.. image:: https://readthedocs.org/projects/semantic-label-creator/badge/?version=latest
        :target: https://semantic-label-creator.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

Python package for converting segmented point cloud to trimmed semantic label masks
as well as converting unsegmented semantic labels into a general point cloud.
The point cloud to semantic conversion is achieved by drawing a circle with fixed diameter
along the line given as points. Additionally the image dataset can be trimmed with or
without padding to indicated size. The semantic to point cloud conversion is done before or
after stitching of images (with or without padding) by selecting at each Z position
all points maxima's and saving their 3D coordinates in array.

* Documentation: https://semantic-label-creator.readthedocs.io/en/latest/

Features
--------
* Convert 3D point cloud from (.am) files to semantic label mask
* Cut and stitch images/masks withe predefined settings
* Convert semantic label mask to 3D point cloud


============
Installation
============


Stable release
--------------

To install Semantic_Label_Creator, run this command in your terminal:

.. code-block:: console

    $ pip install slcpy

This is the preferred method to install Semantic_Label_Creator, as it will always install the most recent stable release.

From sources
------------

The sources for Semantic_Label_Creator can be downloaded from the `Github repo`_.

You can either clone the public repository:

.. code-block:: console

    $ git clone git://github.com/SMLC-NYSBC/Semantic_Label_Creator
    $ python setup.py install

or install is with pip:

.. code-block:: console

    $ pip install slcpy


.. _Github repo: https://github.com/SMLC-NYSBC/Semantic_Label_Creator
.. _tarball: https://github.com/SMLC-NYSBC/Semantic_Label_Creator/tarball/master

=====
Usage
=====

To use Semantic_Label_Creator in a project::

    from slcpy.main import slcpy
    _, label_mask = slcpy_semantic(dir_path,
                                   mask=Ture,
                                   pixel_size=None, 
                                   circle_size=250,
                                   multi_layer=True, 
                                   trim_mask=False, 
                                   trim_size=64)
                                   
    point_could = slcpy_graph(dir_path=label_mask,
                              filter_img=5,
                              down_sampling=Ture)

or with terminal to build semantic label::

    slcpy_semantic -dir C:/... -o C:/.../output 

 string [-dir] Directory of the folder that contain data.
    [-default] os.getcwd() + r'\data'
 string [-o]   Output directory to the folder where all of converted filed are stored.
    [-default] os.getcwd() + r'\data' + r'\output'
 bool   [-m]   Indicate if mask is included.
    [-default] True
 float  [-px]  Images pixel size in Angstrom. If None pixel size is calculated from image metadata
    [-default] None
 int    [-d]   Diameter in Angstrom of a circle that would be drawn a semantic mask
    [-default] 250
 bool   [-l]   Specified if lines should have independent labeling
    [-default] False
 bool   [-t]   Specified if the input image has to be trim to fit labels.
    [-default] True
 int    [-xy]  Define size in pixels of output images.
    [-default] 64
 int    [-z]  Define size in pixels of output images.
    [-default] 64
 bool   [-f]  If True only images containing any data are saved.
    [-default] True
 int    [-s]  Overlay size used for trimming images.
    [-default] 25

with terminal to stitch images::

    slcpy_stitch -dir C:/... -o C:/.../output -m True -pf mask -b True
 string [-dir] Directory of the folder that contain data.
    [-default] os.getcwd() + r'\data'
 string [-o]   Output directory to the folder where all of converted filed are stored.
    [-default] os.getcwd() + r'\data' + r'\output'
 bool   [-m]   If True output images are treated as mask not images.
    [-default] True
 string [-pf]  Additional prefix name for each image.
    [-default] None
 bool   [-b]   If True output stitched image as binary mask.
    [-default] True
 string [-dt]   Output numpy data type.
    [-default] int8

with terminal to build point cloud from image::

    slcpy_graph -dir C:/... -o C:/.../output -f 6 -c 3 -s cvs 
 string [-dir] Directory of the folder that contain data.
    [-default] os.getcwd() + r'\data'
 string [-o]   Output directory to the folder where all of converted filed are stored.
    [-default] os.getcwd() + r'\data' + r'\output'
 int    [-f]   Filter size matrix for denoising.
    [-default] 6
 bool   [-d]   Down-sample point cloud by the factor of.
    [-default] True
 string [-s]   Define format of output point cloud.
    [-default] all
    [-option] all, csv, numpy


HISTORY.rst
=======
History
=======

0.3.6 (2022-01-13)
-------------------
* Include Open3d library for point downsampling
* Maintenance update

0.3.5 (2022-01-03)
-------------------
* HotFix for folder selection on MacOS
* Bugfix in graph creation
* Fix patch creation in Z -> not cutting patches in Z for given size
* Added garbage collection in build graph to manage big files

0.3.4 (2021-11-28)
-------------------
* Added point downsamling 
* Bug fix for calculating patch size when z dimension was bigger then image size

0.3.3 (2021-11-28)
-------------------
* Quick changes for 0.3.3 for build graph tool.
  * Added image output after denoising as an option

0.3.3 (2021-11-28)
-------------------
* Added check for spindletorch to select dtype of output from
  slcpy_stitch
  
0.3.2 (2021-11-19)
-------------------
* Bugfixes for version 0.3.1
* Small update to adapt slcpy to spindletorch and ML packages

0.3.1 (2021-10-31)
-------------------
* Maintained update

0.3.0 (2021-10-12)
-------------------
* Remove pixel size calculation interchange to extracting information from (.am) files
* Added module for extracting 3D coordinates for point cloud from semantic label mask
* Added trimming of whole image with stride to calculate overlaying area and
  number or patches in xyz
* Added module to stitch images into one images based on file names (e.g 1_1_1, where
  each number indicates xyz patch position

0.2.1 (2021-08-13)
-------------------
* Output data are now split to imgs and mask folder if -xy != None

0.2.0 (2021-08-13)
--------------------
* Added trimming of data for multiple images with specific size + recycling of empty labels

0.1.9 (2021-07-17)
-------------------
* Added trimming of input data to label mask size
* Added multiple labels

0.1.8 (2021-07-17)
-------------------
* Bugfix for drawing horizontal lines

0.1.7 (2021-07-09)
-------------------
* Maintenance update

0.1.6 (2021-07-07)
-------------------
* Bugfixes for version 0.1.0
* Setup PyPI and conda

0.1.0 (2021-06-30)
-------------------
* Standardized project entry with cookecutter
* Set up loading of standard data types *.tiff, *.am
* Transform Amira coordinates into pixel value
* Interpolate pixel coordinates and build semantic label


slcpy/utils/trim.py
import math
from os.path import join
from typing import Optional

import numpy as np
from tifffile import tifffile


def trim_images(image: np.ndarray,
                trim_size_xy: int,
                trim_size_z: int,
                multi_layer: bool,
                output: str,
                image_counter: int,
                label_mask: Optional[np.ndarray] = None):
    """
    MODULE FOR TRIMMING DATA TO SPECIFIED SIZES

    Args:
        image: corresponding image for the labels
        label_mask: empty label mask or None if image mask is not included
        trim_size_xy: size of trimming in xy dimension
        trim_size_z: size of trimming in z dimension
        multi_layer: single, or unique value for each lines
        output: name of the output directory for saving
        image_counter: number id of image
    """
    idx = image_counter + 1

    if multi_layer:
        nz, ny, nx, nc = label_mask.shape
    elif multi_layer is False and label_mask is not None:
        nz, ny, nx = label_mask.shape
        nc = None
    else:
        nz, ny, nx = image.shape
        nc = None

    x_axis, y_axis = nx // trim_size_xy, ny // trim_size_xy
    nz_axis = nz // 2

    ny_start, ny_end = -trim_size_xy, 0
    nz_start = nz_axis - (trim_size_z // 2)
    nz_end = nz_axis + (trim_size_z // 2)
    if nz_start < 0:
        nz_start = 0
        nz_end = nz

    if (nz_end - nz_start) != trim_size_z:
        padding = True
    else:
        padding = False

    for i in range(1, y_axis + 1):
        ny_start += trim_size_xy
        ny_end += trim_size_xy
        nx_start, nx_end = -trim_size_xy, 0

        for j in range(1, x_axis + 1):
            nx_start += trim_size_xy
            nx_end += trim_size_xy
            img_name = str(idx) + '.tif'
            mask_name = str(idx) + '_mask.tif'
            trim_img = np.zeros((trim_size_z,
                                 trim_size_xy,
                                 trim_size_xy))

            if not padding:
                trim_img = image[nz_start:nz_end,
                                 ny_start:ny_end,
                                 nx_start:nx_end]
            else:
                trim_img[0:(nz_end - nz_start),
                         0:(ny_end - ny_start),
                         0:(nx_end - nx_start)] = image[nz_start:nz_end,
                                                        ny_start:ny_end,
                                                        nx_start:nx_end]

            if label_mask is not None:
                if nc is None:
                    trim_mk = np.zeros((trim_size_z,
                                        trim_size_xy,
                                        trim_size_xy))

                    if not padding:
                        trim_mk = label_mask[nz_start:nz_end,
                                             ny_start:ny_end,
                                             nx_start:nx_end]
                    else:
                        trim_mk[0:(nz_end - nz_start),
                                0:(ny_end - ny_start),
                                0:(nx_end - nx_start)] = label_mask[nz_start:nz_end,
                                                                    ny_start:ny_end,
                                                                    nx_start:nx_end]
                else:
                    trim_mk = np.zeros((trim_size_z,
                                        trim_size_xy,
                                        trim_size_xy,
                                        nc))
                    if not padding:
                        trim_mk = label_mask[nz_start:nz_end,
                                             ny_start:ny_end,
                                             nx_start:nx_end,
                                             :]
                    else:
                        trim_mk[0:(nz_end - nz_start),
                                0:(ny_end - ny_start),
                                0:(nx_end - nx_start),
                                :] = label_mask[nz_start:nz_end,
                                                ny_start:ny_end,
                                                nx_start:nx_end,
                                                :]

                    trim_mk = label_mask[nz_start:nz_end,
                                         ny_start:ny_end,
                                         nx_start:nx_end,
                                         :]

            if np.all(trim_mk[:, :, :] == 0):
                idx = idx
            else:
                # Hard transform between int8 and uint8
                if np.min(trim_img) < 0:
                    trim_img = trim_img + 128

                tifffile.imwrite(join(output, 'mask', mask_name),
                                 np.array(trim_mk, 'int8'))
                tifffile.imwrite(join(output, 'imgs', img_name),
                                 np.array(trim_img, 'int8'))
                idx += 1

    return idx


def trim_to_patches(image: np.ndarray,
                    trim_size_xy: int,
                    trim_size_z: int,
                    multi_layer: bool,
                    output: str,
                    image_counter: int,
                    label_mask: Optional[np.ndarray] = None,
                    stride=25):
    """
    FUNCTION TO TRIMMED IMAGE AND MASKS TO SPECIFIED SIZE

    Output images are saved as tiff with naming shame 1_1_1_25. Where
    number indicate grid position in xyz. Last number indicate stride.

    Args:
        image: Corresponding image for the labels
        label_mask: Empty label mask
        trim_size_xy: Size of trimming in xy dimension
        trim_size_z: Size of trimming in z dimension
        multi_layer: Single, or unique value for each lines
        output: Name of the output directory for saving
        image_counter:
        stride: Trimming step size

    Returns:
        Saved trimmed images as tiff in specified folder
    """
    idx = image_counter

    if multi_layer:
        nz, ny, nx, nc = label_mask.shape
    elif multi_layer is False and label_mask is not None:
        nz, ny, nx = label_mask.shape
        nc = None
    else:
        nz, ny, nx = image.shape
        nc = None

    if trim_size_xy is not None or trim_size_z is not None:
        assert nx >= trim_size_xy, \
            "trim_size_z should be equal or greater then X dimension!"
        assert ny >= trim_size_xy, \
            "trim_size_z should be equal or greater then Y dimension!"
    else:
        assert stride is not None, \
            "Trim sizes or stride has to be indicated!"
        trim_size_xy = 64
        trim_size_z = 64

    # Calculate number of patches, patch sizes, and stride for xyz
    x, y, z = math.ceil(nx / trim_size_xy), \
        math.ceil(ny / trim_size_xy), \
        math.ceil(nz / trim_size_z)

    x_padding, y_padding, z_padding = (trim_size_xy + ((trim_size_xy - stride) * (x - 1))) - nx, \
                                      (trim_size_xy + ((trim_size_xy - stride) * (y - 1))) - ny, \
                                      (trim_size_z + ((trim_size_z - stride) * (z - 1))) - nz

    # Adapt number of patches for trimming
    if trim_size_xy is not None or trim_size_z is not None:
        while x_padding < 0:
            x += 1
            x_padding += trim_size_xy - stride
        while y_padding < 0:
            y += 1
            y_padding += trim_size_xy - stride
        while z_padding < 0:
            z += 1
            z_padding += trim_size_z - stride

    # Adapt patch size for trimming
    else:
        while x_padding <= 0 or y_padding <= 0:
            trim_size_xy += 1
            x_padding = (trim_size_xy +
                         ((trim_size_xy - stride) * (x - 1))) - nx
            y_padding = (trim_size_xy +
                         ((trim_size_xy - stride) * (y - 1))) - ny

        while z_padding < 0:
            trim_size_z += 1
            z_padding = (trim_size_z + ((trim_size_z - stride) * (z - 1))) - nz

    # Expand image of a patch
    image_padded = np.pad(image,
                          [(0, z_padding), (0, y_padding), (0, x_padding)],
                          mode='constant')

    if label_mask is not None:
        if nc is None:
            mask_padded = np.pad(label_mask,
                                 [(0, z_padding), (0, y_padding), (0, x_padding)],
                                 mode='constant')
        else:
            mask_padded = np.pad(label_mask,
                                 [(0, z_padding), (0, y_padding),
                                  (0, x_padding), (0, 0)],
                                 mode='constant')

    # Trim image and mask with stride
    z_start, z_stop = 0 - (trim_size_z - stride), 0

    for i in range(z):
        z_start = z_start + trim_size_z - stride
        z_stop = z_start + trim_size_z
        y_start, y_stop = 0 - (trim_size_xy - stride), 0

        for j in range(y):
            y_start = y_start + trim_size_xy - stride
            y_stop = y_start + trim_size_xy
            x_start, x_stop = 0 - (trim_size_xy - stride), 0

            for k in range(x):
                x_start = x_start + trim_size_xy - stride
                x_stop = x_start + trim_size_xy

                img_name = str("{}_{}_{}_{}_{}.tif".format(
                    idx, k, j, i, stride))
                mask_name = str(
                    "{}_{}_{}_{}_{}_mask.tif".format(idx, k, j, i, stride))

                trim_img = image_padded[z_start:z_stop,
                                        y_start:y_stop,
                                        x_start:x_stop]

                if label_mask is not None:
                    if nc is None:
                        trim_mk = mask_padded[z_start:z_stop,
                                              y_start:y_stop,
                                              x_start:x_stop]
                    else:
                        trim_mk = mask_padded[z_start:z_stop,
                                              y_start:y_stop,
                                              x_start:x_stop,
                                              :]
                    tifffile.imwrite(join(output, 'mask', mask_name),
                                     np.array(trim_mk, 'int8'))

                if label_mask is not None:
                    tifffile.imwrite(join(output, 'imgs', img_name),
                                     np.array(trim_img, 'int8'))
                else:
                    tifffile.imwrite(join(output, img_name),
                                     np.array(trim_img, 'int8'))
    idx += 1
    return idx
